src/Symfony/Component/Encryption/SymfonyEncryptionToken.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption;
+
+use Symfony\Component\Encryption\Exception\DecryptionException;
+use Symfony\Component\Encryption\Exception\MalformedCipherException;
+
+/**
+ * This class is responsible for creating and parsing Symfony Encryption Tokens.
+ * It is an helper class for the EncryptionInterface implementations. In all common
+ * scenarios, one does not need to interact with this class directly.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class SymfonyEncryptionToken implements \Stringable
+{
+    /**
+     * Algorithm used to encrypt the message.
+     */
+    private string $algorithm;
+    private string $version;
+    private string $ciphertext;
+
+    /**
+     * @var array<string, string>
+     */
+    private array $headers = [];
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * @param array<string, string> $headers keys MUST be ascii values only
+     */
+    public static function create(string $algorithm, string $ciphertext, array $headers = []): self
+    {
+        $model = new self();
+        $model->algorithm = $algorithm;
+        $model->ciphertext = $ciphertext;
+        $model->headers = $headers;
+
+        return $model;
+    }
+
+    /**
+     * Take a string representation of the ciphertext and parse it into an object.
+     *
+     * @throws MalformedCipherException
+     */
+    public static function parse(string $input): self
+    {
+        $parts = explode('.', $input);
+        if (!\is_array($parts) || 3 !== \count($parts)) {
+            throw new MalformedCipherException();
+        }
+
+        [$headersString, $ciphertext, $hashSignature] = $parts;
+
+        $headersString = self::base64UrlDecode($headersString);
+        $ciphertext = self::base64UrlDecode($ciphertext);
+        $hashSignature = self::base64UrlDecode($hashSignature);
+
+        // Check if data has been modified
+        $hash = hash('sha256', $headersString.$ciphertext);
+        if (!hash_equals($hash, $hashSignature)) {
```

Comment:
Misnomer: `$hashSignature` isn't, [at all](https://paragonie.com/blog/2015/08/you-wouldnt-base64-a-password-cryptography-decoded), a hash signature.

If you were doing the right thing, it would be a MAC, not a signature, but this just a SHA256 hash *without* any secrets. This is trivially bypassed: Just calculate `hash('sha256', $headersString.$ciphertext)` for any arbitrary data you want to feed in.

You're probably better off calling this a checksum, or equivalent, so as to not over-sell the security properties you're shooting for.

Reply:
This setup also looks susceptible to [canonicalization attacks](https://soatok.blog/2021/07/30/canonicalization-attacks-against-macs-and-signatures/). If the intent here is to have something like HMAC, you'll want to also revisit this.